cache-filter/src/unique_callout.rs

```diff
@@ -129,30 +170,171 @@ pub fn free_callout_lock(
+        queue_id: queue_id.unwrap(),
+        http_context_id: context.context_id,
+    };
+    info!(
+        context.context_id,
+        "thread({}): trying to add to callout waitlist", context.root_id
+    );
+
+    // Unless there is some internal error, there is a guarantee that 1 thread successfully
+    // writes to the cache. So, keep trying and eventually every thread will win.
+    loop {
+        match get_shared_data(&waiters_key) {
+            Ok((Some(bytes), Some(cas))) => {
+                info!(
+                    context.context_id,
+                    "thread({}): adding to existing waitlist", context.root_id
+                );
+                match bincode::deserialize::<Vec<CalloutWaiter>>(&bytes) {
+                    Ok(mut callout_waiters) => {
+                        callout_waiters.push(callout_waiter.clone());
+                        let serialized_cw =
+                            match bincode::serialize::<Vec<CalloutWaiter>>(&callout_waiters) {
+                                Ok(res) => res,
+                                Err(e) => {
+                                    return Err(UniqueCalloutError::SerializeFail {
+                                        what: "Vec<CalloutWaiter>",
+                                        reason: *e,
+                                    })
+                                }
+                            };
+                        if let Err(Status::CasMismatch) =
+                            set_shared_data(&waiters_key, Some(&serialized_cw), Some(cas))
+                        {
+                            continue;
+                        }
+                        break;
+                    }
+                    Err(e) => {
+                        return Err(UniqueCalloutError::DeserializeFail {
+                            what: "Vec<CalloutWaiter>",
+                            reason: *e,
+                        });
+                    }
+                }
+            }
+            Ok((_, cas)) => {
+                info!(
+                    context.context_id,
+                    "thread({}): creating a new waitlist", context.root_id
+                );
+                let serialized_cw =
+                    match bincode::serialize::<Vec<CalloutWaiter>>(&vec![callout_waiter.clone()]) {
+                        Ok(res) => res,
+                        Err(e) => {
+                            return Err(UniqueCalloutError::SerializeFail {
+                                what: "Vec<CalloutWaiter>",
+                                reason: *e,
+                            })
+                        }
+                    };
+                // Again exploiting host implementation to avoid multi-initialization:
+                let cas = cas.unwrap_or(1);
+                if let Err(Status::CasMismatch) =
+                    set_shared_data(&waiters_key, Some(&serialized_cw), Some(cas))
+                {
+                    continue;
```

Comment:
It's nice to add a log here so we know `CasMismatch` happened.

Reply:
Added